Classync - Simon Frisk - Final Project

simfrisk opened the pull request:
‼️‼️ Major update just made, I will update the PR description shortly ‼️‼️

**Pre made users**

**Teacher user:**
Username: technigo@teacher.com
Password: technigo

**Student user**
technigo@student.com
Password: technigo

____________________________________________________


Quick video recap if you don't like to read😂

👇👇👇👇👇👇👇👇👇
https://github.com/user-attachments/assets/72eaf53e-e7be-4a79-aaa3-c7936c73c4e9
👆👆👆👆👆👆👆👆👆

❓ What it is:

A video review platform that solves the problem of giving specific feedback on video content. Instead of saying "around the 3-minute mark" teachers and students can comment directly on video timestamps and have meaningful discussions.

🔗 Links

Netlify link:
https://class-review.netlify.app/

Tips: 
- you don't have to add a profile image to sign up
- Make sure to try teacher and student users
- Watch my video for a quick recap of the project.

Backend documentation: 
https://project-final-simon.onrender.com/

✨ Core Features

- Time-Coded Video Comments - Click anywhere on the video timeline to add comments at specific moments.
- Interactive Video Player - Custom video controls with timeline markers, volume controls, and full-screen support.
- Comment System - Three types: questions, public comments, and private notes with nested replies and likes.
- Class & Project Management - Organize educational content by classes with teacher/student role separation.
- Teacher Dashboard - Track unchecked student questions and manage content across classes with direct connection to the timeline on click.
- Workspaces - Like an organization, school or such
- Teams - This is like a school class, team or department with the possibility to add students and teacher to get different access to different classes to be able to access different vidoes.

🛠️ Tech Stack

- Frontend: React 19 + TypeScript, Styled Components, Framer Motion, Zustand state management.
- Backend: Node.js + Express, MongoDB with Mongoose, JWT authentication.
- Media: Cloudinary for video/photo storage and automatic thumbnail generation.
- API: RESTful endpoints with comprehensive Swagger documentation:
- State Management: Separate Zustand stores for projects, comments, users, and video controls.

🔮 Future Plans

- The users page is a recent addition more a taste of coming features. Planning to add advanced user management, real-time notifications, and enhanced collaboration features as bonus enhancements.
- I would also want to add Web Socket for real time comment fetching but to scared to add it before the demos.

🤔 Things to know

- I know now that naming the courses as class or classes in the code as a bad idé but so much to refactor for now.
- My goal with TypeScript was just to use it so I used a bit more AI on it then I would have liked but it learned a lot!
- The touch areas on the timeline markers does have a bigger touch area then Lighthouse says but it does work so the score is not correctly reflected.


```
  &::before {
    content: "";
    position: absolute;
    top: 50%;
    left: 50%;
    border-radius: 50%;
    transform: translate(-50%, -50%);
    width: 22px;
    height: 22px;
    background: transparent;
    cursor: pointer;

https://github.com/user-attachments/assets/08be8dcd-28d8-453b-b741-0e6a55dbc8e3


  }
```

Hope you enjoy the site! 😊

/Simon

simfrisk committed:
Fixed clicking on one comment!

simfrisk committed:
added backend like endpoint

simfrisk committed:
forgot to build

simfrisk committed:
made a default for the model likes

simfrisk committed:
Got like frontend to work

simfrisk committed:
Added likes in the backend to reply

simfrisk committed:
Added some styling

simfrisk committed:
Added enpoint route

simfrisk committed:
Small fix

simfrisk committed:
Fixed the dependencies

simfrisk committed:
Bug fix

simfrisk committed:
Got the patch to work!

simfrisk committed:
added some backend populate

simfrisk committed:
testing a fix

simfrisk committed:
Adde populate endpoint patcgh

simfrisk committed:
Added .en files with base url

simfrisk committed:
Added more baseurl

simfrisk committed:
Added all api basepath

simfrisk committed:
Added some styling

simfrisk committed:
Added more styling

simfrisk committed:
Did some more styling for darkmode

simfrisk committed:
Small theme

simfrisk committed:
Added REgions to the stores

simfrisk committed:
Added card Header

simfrisk committed:
Added the main

simfrisk committed:
Added the footer

simfrisk committed:
Added more cleanup in reply card

simfrisk committed:
Added toggle

simfrisk committed:
Added Checks to checkboxes

simfrisk committed:
Fixed the css for the teachers cards

simfrisk committed:
Fixed the scaleing problem to the review page

simfrisk committed:
Bug fix css review sizing

simfrisk committed:
Added profile image to nav

simfrisk committed:
Small profile resizing

simfrisk committed:
Made the lanfing page newer

simfrisk committed:
Added more styling to landing cards

simfrisk committed:
Small color tweeks

simfrisk committed:
Forgot typescritpt

simfrisk committed:
Added documentation

simfrisk committed:
add changes

simfrisk committed:
build

simfrisk committed:
Added swagger!

simfrisk committed:
Added some envs and so

simfrisk committed:
Made a src folder

simfrisk committed:
Added routes

simfrisk committed:
Added a video to start

simfrisk committed:
Added alot to the landing page

simfrisk committed:
Updated landing page!

simfrisk committed:
Small backend rout fix

simfrisk committed:
Added styling

simfrisk committed:
Added some fixes

simfrisk committed:
More darkmode

simfrisk committed:
Darkmode finished for now

simfrisk committed:
Rearange some theme files

simfrisk committed:
Added spaing and stuff with new queriees

simfrisk committed:
Add back

simfrisk committed:
I can now confirm delete project

simfrisk committed:
Added some styling

simfrisk committed:
More styling fixes

simfrisk committed:
Added delete class ui and function

simfrisk committed:
Added faster video load

simfrisk committed:
Added some more stuff

simfrisk committed:
Some refactoring of box function

simfrisk committed:
Added some styling to teachers page

simfrisk committed:
Styling to select

simfrisk committed:
Some more logic and styling to class handeling

simfrisk committed:
centered are you shure

simfrisk committed:
Fixed so that classes hide if 0 on teachers page

simfrisk committed:
Rty to fix

simfrisk committed:
Change the ts config file

simfrisk committed:
Tyed server in src

simfrisk committed:
Added patch projects

simfrisk committed:
Added swagger uppdates

simfrisk committed:
Added headers to swagger

simfrisk committed:
Added soe changes to swagger

simfrisk committed:
Try changing middleware

simfrisk committed:
Added swagger stuff

simfrisk committed:
Added edit to project

simfrisk committed:
added is checked check

simfrisk committed:
Ts fix

simfrisk committed:
Made the props for the teachers page

simfrisk committed:
Tried fixing the teachers page

simfrisk committed:
Test wit teachers

simfrisk committed:
Redid the teachers logic

simfrisk committed:
Bug type fix

simfrisk committed:
Naming issue

simfrisk committed:
Made the delete box hide the edit box

simfrisk committed:
Added patch class to backend

simfrisk committed:
Got the patch class to work even though it is a bit ugly

simfrisk committed:
Small class tweek

simfrisk committed:
Small spacing fix

simfrisk committed:
Cleaned up the project a bit

simfrisk committed:
Added mobile delete class

simfrisk committed:
Dakrmode to buttons

simfrisk committed:
Added model created by

simfrisk committed:
Added controller Creadedby project

simfrisk committed:
Added creaded by

simfrisk committed:
Added created by project to frontend

simfrisk committed:
Added styling fix

simfrisk committed:
Added the project image back to teachers page

simfrisk committed:
Reforamated the code with prettier

simfrisk committed:
Small design fixes

simfrisk committed:
Fixed backround is etiting

simfrisk committed:
Made user not unique

simfrisk committed:
Removed unieq name

simfrisk committed:
Added delete user

simfrisk committed:
added placeholder user

simfrisk committed:
Fixed rout path

simfrisk committed:
Adde password button and some error handeling for the user

simfrisk committed:
Added area and stuff the the navigation

simfrisk committed:
Cleaned up the navigation

simfrisk committed:
Fixed route path

simfrisk committed:
Added better error handeling in login and signup

simfrisk committed:
Added more error handeling

simfrisk committed:
Added max 100mb file size

simfrisk committed:
Cleaned up the login page

simfrisk committed:
SignUp page aria and stuff

simfrisk committed:
Added some aria to ProjectLib

simfrisk committed:
Added regions

simfrisk committed:
Added more aria to library

simfrisk committed:
Small bug style

simfrisk committed:
Footer link fix

simfrisk committed:
added backend to project teacher

simfrisk committed:
Added more aria

simfrisk committed:
Added more aria to commentsection

simfrisk committed:
Added button navigation

simfrisk committed:
Adde more aria

simfrisk committed:
More aria

simfrisk committed:
Last aria fixes

simfrisk committed:
fixed type bugs

simfrisk committed:
Bug fixes

simfrisk committed:
Added some video and text

simfrisk committed:
Small update sizing

simfrisk committed:
Styling fixes

simfrisk committed:
Landing page fix

simfrisk committed:
Added a styling and function to Target slides

simfrisk committed:
Add some styling

simfrisk committed:
Update README.md with detailed project description, features, technologies used, and setup instructions for Classync platform

simfrisk committed:
Added the feature section better

simfrisk committed:
Added sizing and stuff

simfrisk committed:
Refactor SubTitle component to use h3 and adjust styling for improved layout

simfrisk committed:
More sizing

simfrisk committed:
Refactor layout and styling for footer and feature titles; update demo text for clarity

simfrisk committed:
Refactor CommentForm and TeachersPage components for improved code readability and consistency; adjust button padding and border radius for better styling

simfrisk committed:
Update styling across various components for improved consistency and readability; adjust colors, padding, and background properties to align with theme settings.

simfrisk committed:
Update feature card styles to use specific color values for background and text, enhancing visual consistency.

simfrisk committed:
Add headings and improve styling in CreateClass and CreateProject components; adjust padding and layout for better user experience

simfrisk committed:
Refactor styling in various components to enhance theme consistency; update background colors, hover effects, and remove unnecessary styles for improved readability.

simfrisk committed:
Add fullscreen functionality to VideoSection component; implement fullscreen toggle and update controls layout for improved user experience.

simfrisk committed:
Refactor VideoSection component to simplify fullscreen handling; streamline controls layout and enhance responsiveness based on fullscreen state.

simfrisk committed:
Enhance accessibility and semantics in navigation and landing page components; add ARIA roles and labels, improve structure for better screen reader support, and refine styling for consistency.

simfrisk committed:
Add secondary color to dark theme; enhance color palette for improved design consistency.

simfrisk committed:
Refactor accessibility roles in navigation and landing page components; remove unnecessary ARIA roles to enhance semantic structure and improve screen reader support.

simfrisk committed:
Update image formats across multiple components to WebP for improved performance; replace PNG and JPEG images with WebP versions in landing, login, and signup pages.

simfrisk committed:
Refactor HTML structure and improve accessibility in navigation components; update page titles, enhance ARIA attributes, and streamline layout for better user experience.

simfrisk committed:
Refactor FeatureSection component by removing unnecessary ARIA attributes from CardWrapper for improved semantic structure and accessibility.

simfrisk committed:
Refactor accessibility roles in landing page components by removing unnecessary ARIA roles from Demos, FeatureCards, and Target components to enhance semantic structure and improve screen reader support.

simfrisk committed:
Remove Beta tag and VideoBackground component from Navigation and Hero sections respectively to streamline code and improve component clarity.

simfrisk committed:
Change SubTitle component from h3 to h2 for improved semantic structure and accessibility in Hero section.

simfrisk committed:
Refactor CommentHeader, CommentSection, CommentCardFooter, and VideoSection components to improve accessibility and styling; remove unnecessary ARIA attributes, adjust padding and margins, and enhance hover effects for better user experience.

simfrisk committed:
Refactor LoginForm, LoginHeader, SignUpForm, SignUpHeader, VideoLibraryPage, ReviewPage, and related components to improve accessibility and styling; consolidate image handling, enhance semantic structure, and streamline code for better user experience.

simfrisk committed:
Refactor Section component props to use dollar sign prefix for styled-components; update related components in landing page and navigation for improved consistency and clarity.

simfrisk committed:
Enhance AFeatureCard component by adding description prop for improved content display; update feature cards with relevant descriptions for better user context. Adjust styling in various components for improved layout and accessibility.

simfrisk committed:
Update SignUpForm styling for radio buttons and enhance TeacherProjectCard text display with ellipsis for improved readability and layout consistency.

simfrisk committed:
Enhance UserSchema by adding lowercase and trim options to the email field for improved data consistency and validation.

simfrisk committed:
Enhance various models and components by adding maxlength and trim options to string fields for improved data validation and consistency. Update frontend forms to reflect these changes, ensuring better user input handling.

simfrisk committed:
Refactor routes in backend to improve organization and clarity; add new endpoints for classes, projects, comments, replies, likes, and user authentication. Update documentation comments for better structure and readability.

simfrisk committed:
Update documentation comment in routes file to clarify purpose, changing from 'DOCUMENTATION' to 'ALTERNATIVE DOCUMENTATION' for improved clarity.

simfrisk committed:
Add new endpoint for retrieving users in routes; include getUsers controller and update documentation comment for clarity.

simfrisk committed:
Enhance user management by updating getUsers controller to exclude sensitive fields, adding UserPage route in frontend, and implementing getAllUsers function in user store for fetching user data.

simfrisk committed:
Refactor UserPage to display users by role, separating teachers and students into distinct sections. Implement sorting functionality in user store to organize users based on role priority.

simfrisk committed:
Enhance responsive design across components by adding media queries for small phone sizes. Update CreateProject and Project components to improve layout adaptability, and refine CommentHeader button group styling for better alignment and spacing on smaller screens.

simfrisk committed:
Restrict edit and delete options in Project component to users with teacher role, enhancing access control based on user permissions.

simfrisk committed:
Refactor deleteUser controller to streamline user ID initialization and update TeacherProjectCard component to filter comments by student role, improving comment visibility and management.

simfrisk committed:
Refactor UserPage to improve layout and accessibility by adding a navigation component, restructuring user display into distinct sections for teachers and students, and enhancing styling for better visual clarity and responsiveness.

simfrisk committed:
Refactor navigation components to use Link for UserProfile, enhancing routing functionality. Add hover effect for better user interaction. Update UserPage to display current user information with improved layout and accessibility features, including distinct sections for teachers and students.

simfrisk committed:
Refactor UserPage styling for improved responsiveness and visual consistency. Adjust padding, margins, and font sizes across various components, ensuring better layout adaptability on different screen sizes. Enhance user card design and section headers for a more cohesive user experience.

simfrisk committed:
Update getUsers controller to exclude sensitive fields from the response, enhancing user data privacy and security.

simfrisk committed:
Update TeachersPage to filter projects by student role for unchecked questions, improving data accuracy in project management.

simfrisk committed:
Refactor TeachersPage to filter unchecked comments specifically by student role, enhancing the accuracy of class selection for project management.

simfrisk committed:
Enhance TeachersPage to ensure only student-created unchecked comments are displayed, improving the relevance of project feedback and visibility.

simfrisk committed:
Refactor TeachersPage to simplify comment filtering by removing unnecessary checks for student role, streamlining the display of unchecked question comments.

simfrisk committed:
Update TeacherProjectCard styling for improved layout consistency. Adjust button gap to column-gap and modify font size in CommentFooter for better readability.

simfrisk committed:
Refactor TeachersPage to replace SmallButton with styled WorkingButton and ActionButton components, enhancing button styling and interactivity. Adjust PageTitle margin for improved layout consistency.

simfrisk committed:
Enhance TeachersPage styling by adjusting margins and padding for ClassSelect, WorkingButton, and ActionButton components. Improve responsiveness with media queries for better layout on different screen sizes.

simfrisk committed:
Add video source to ExplainerCard in demo component for enhanced content delivery

simfrisk committed:
Implement user deletion functionality in UserPage with confirmation modal. Add editing capabilities for teachers and students, enhancing user management experience. Update user store to support deleteUser action and improve user data handling.

simfrisk committed:
Filter out current user from teachers list in UserPage and adjust student count display to exclude placeholder user.

simfrisk committed:
Add react-icons dependency and refactor SignUpForm layout in UserPage

- Added `react-icons` package to package.json for icon support.
- Refactored SignUpForm component layout for improved structure and accessibility.
- Updated UserPage to include SignUpForm for user creation functionality.
- Adjusted styling for better visual consistency and user experience.

simfrisk committed:
Update README and UserPage to enhance user creation functionality

- Removed outdated planning details from README for clarity.
- Updated UserPage to conditionally display the Create User section based on the current user's role, indicating that the feature is in progress.

simfrisk committed:
Refactor user deletion handling in UserPage and improve error handling in userStore

- Introduced a new handleDeleteUser function in UserPage to streamline user deletion logic.
- Updated DeleteUserButton onClick handlers to utilize the new function for better readability.
- Changed the Create User section header to indicate it is in development.
- Modified userStore to handle optional _id for AuthUser and improved error handling for user-related actions.

simfrisk committed:
Add loading state to VideoSection with pulsing background and spinning circle

- Implemented a loading state in VideoSection to enhance user experience while the video is loading.
- Added PulsingBackground and SpinningCircle styled components for visual feedback during loading.
- Updated the render logic to conditionally display the loading state based on videoLoaded status.

simfrisk committed:
Add user patch functionality and update user role schema

- Introduced a new PATCH endpoint for updating user details in routes.ts.
- Enhanced UserSchema in user.ts to include an enum for user roles, allowing only "teacher" or "student" as valid options.

simfrisk committed:
Refactor UserPage to streamline user management sections and enhance code organization

- Removed redundant state management and user deletion logic from UserPage.
- Introduced separate components for CurrentUserSection, TeachersSection, StudentsSection, and CreateUserSection for better modularity.
- Updated userStore to include a new updateUser function for modifying user details.

simfrisk committed:
Add editing functionality to CurrentUserSection for user details management

- Introduced state management for editing user information, including name, email, role, and profile image.
- Added buttons for saving changes and canceling edits, enhancing user interaction.
- Implemented input fields for user data, allowing real-time updates and improved user experience.

simfrisk committed:
Add CreateTeam component and enhance UserPage with tab navigation

- Integrated CreateTeam component into App and UserPage for team management.
- Implemented tab navigation in UserPage to switch between users, teams, and additional options.
- Added styling for action buttons to improve user interaction and experience.

simfrisk committed:
Remove Navigation component from CreateTeam page for a cleaner layout.

simfrisk committed:
Refactor SignUpPage layout and styling for improved user experience

- Updated SignUpPage to use min-height for better responsiveness.
- Adjusted margins in Card and StyledButton components for a cleaner layout.
- Removed logo from SignUpHeader to simplify the design.

simfrisk committed:
Enhance routing and navigation for teacher and admin roles

- Updated App component to include new routes for teachers and admin functionalities, utilizing RequireRole for access control.
- Modified Desktop and Mobile Navigation components to reflect the new route paths for teachers and admin users.
- Improved comments for clarity on public and protected routes.

simfrisk committed:
Add TeamsSection to UserPage and update profile image handling

- Integrated TeamsSection into UserPage for improved team management.
- Updated profile image fallback from default-avatar.png to logo2.png in CurrentUserSection, StudentsSection, and TeachersSection for consistency.

simfrisk committed:
Enhance data models with schemaVersion and additional fields

- Added schemaVersion field with default value "v2" to Class, Comment, Project, Reply, and User models for versioning.
- Introduced workspaceId to Class model and expanded User model with workspaces, teams, and assignedTeams fields for better data organization.

simfrisk committed:
Implement user role management and enhance user details editing

- Added functionality to manage user roles within the application, allowing for role assignments and updates.
- Improved the user details editing process by refining input handling and validation for a smoother user experience.
- Updated relevant components to reflect changes in user role management and ensure consistency across the application.

simfrisk committed:
Add workspace and team management routes to backend

simfrisk committed:
Refactor routes and enhance class management functionality

- Updated routes to improve organization and clarity, including restructuring team routes under specific workspaces.
- Enhanced class retrieval and creation by adding workspace filtering and ensuring workspace ID is required for class creation.
- Updated class model to include workspace ID for better data association.
- Improved API documentation to reflect new query parameters for class retrieval.

simfrisk committed:
Update routes and controllers to enforce workspace context for classes and teams

- Refactored routes to require workspace ID for class and team operations, enhancing data organization.
- Updated controller logic to retrieve and manage classes and teams within the specified workspace.
- Improved API documentation to reflect changes in endpoint structure and required parameters.

simfrisk committed:
Refactor team management routes and enhance postTeam functionality

- Updated team management routes to include user IDs in the path for adding members, teachers, and classes, improving clarity and specificity.
- Enhanced postTeam controller to accept arrays of members, teachers, and classes, allowing for bulk assignments during team creation.
- Improved API documentation to reflect changes in endpoint structure and request body requirements.

simfrisk committed:
Fix optional chaining for user ID in postWorkspace controller

- Updated the postWorkspace controller to use optional chaining for accessing the user ID, ensuring robustness against undefined user objects.
- This change improves error handling and prevents potential runtime errors when the user is not authenticated.

simfrisk committed:
Enhance workspace management and user flow

- Updated postWorkspace controller to automatically add the creator to the newly created workspace, improving user experience by streamlining workspace association.
- Added a new route for creating workspaces in the App component, allowing users to navigate directly to the workspace creation page.
- Refactored sign-up form submission to automatically set the user role as "teacher" and redirect to the create workspace page upon successful sign-up.
- Enhanced class and user stores to support workspace associations, including methods for fetching classes by workspace and setting user workspace, improving data organization and user context.

simfrisk committed:
Implement workspace and team management enhancements

- Added new routes for workspace and team management, including creation, retrieval, and deletion of workspaces and teams, ensuring operations are tied to specific workspaces.
- Updated class management routes to require workspace context, allowing for better organization and retrieval of classes within a workspace.
- Enhanced class model to include workspace ID, improving data association and integrity.
- Improved error handling in controllers to ensure robustness against undefined user objects and provide clearer error messages.
- Updated API documentation to reflect changes in endpoint structures and required parameters.

simfrisk committed:
Add user workspaces functionality and enhance navigation components

- Introduced a new route for fetching user-specific workspaces, improving user experience by allowing users to view their associated workspaces.
- Updated Desktop and Mobile navigation components to include a workspace selector, enabling users to switch between workspaces seamlessly.
- Refactored UserProfile component to a button for better accessibility and interaction, enhancing the overall user interface.
- Improved state management for workspace selection in navigation components, ensuring a responsive and dynamic user experience.

simfrisk committed:
Refactor workspace selection and management in navigation and project components

- Updated WorkspaceSelector to utilize currentWorkspaceId from workspace store, enhancing workspace selection logic.
- Modified ProjectsList to load current workspace and fetch classes based on the selected workspace, improving data retrieval.
- Enhanced workspace store with methods for setting and loading the current workspace from local storage, ensuring consistent user experience across components.

simfrisk committed:
Update class fetching logic to require workspace ID

- Modified fetchClasses method in classStore to require workspaceId, ensuring classes are fetched in the context of the current workspace.
- Updated SideMenu and TeachersPage components to utilize currentWorkspaceId when fetching classes, improving data retrieval and user experience.

simfrisk committed:
Enhance workspace integration in class management

- Updated deleteClass, postClass, and getWorkspaceById controllers to manage class associations with workspaces, ensuring classes are properly linked and removed from the workspace context.
- Modified Workspace model to include a classes array, facilitating better organization of classes within each workspace.
- Improved SideMenu component to fetch user workspaces automatically, enhancing user experience by streamlining workspace selection.

simfrisk committed:
Implement workspace validation in class creation

- Added workspace ID validation in CreateClass component to ensure a workspace is selected before creating a class.
- Updated addClass function to accept workspace ID, enhancing class management within the context of the current workspace.

simfrisk committed:
Add user workspaces retrieval and enhance class management

- Introduced a new route for fetching user-specific workspaces, improving user experience by allowing users to view their associated workspaces.
- Updated class management controllers to ensure classes are properly linked to their respective workspaces, enhancing data integrity.
- Modified Workspace model to include a classes array, facilitating better organization of classes within each workspace.
- Improved error handling in class-related operations to ensure robustness against undefined workspace references.

simfrisk committed:
Refactor class management to improve workspace integration

- Updated class-related controllers to ensure proper linking and management of classes within their respective workspaces.
- Enhanced error handling for class operations to improve robustness against undefined workspace references.
- Modified class fetching logic to require workspace ID, ensuring classes are retrieved in the context of the current workspace.

simfrisk committed:
Enhance user and workspace invitation management

- Added functionality to handle workspace and team invitation links, allowing users to join workspaces and teams via invitation tokens.
- Updated user creation logic to process invitation tokens, automatically associating users with the relevant workspace and team upon sign-up.
- Improved API documentation for invitation-related endpoints, ensuring clarity on usage and expected parameters.
- Enhanced front-end components to validate invitation tokens and provide user feedback during the sign-up process.

simfrisk committed:
Enhance user and invitation management functionality

- Updated postSession and postUser controllers to include user teams in the response, ensuring better user context.
- Improved workspace invitation API documentation to clarify permissions and response structures, emphasizing teacher-only access for certain actions.
- Added new endpoints for managing invitation tokens and retrieving invitation history, enhancing user experience and functionality.
- Refactored front-end components to accommodate role-based access for user pages and team management, ensuring appropriate visibility for teachers and students.

simfrisk committed:
Refactor video library page and enhance class management features

- Updated VideoLibraryPage to include modal for creating classes when no classes are available, improving user experience for teachers.
- Enhanced ChooseClassMessage component to conditionally render a message and button for creating a class if none exist, streamlining class management.
- Modified ProjectsList to display a message prompting class creation for teachers when no classes are available, ensuring clarity in user actions.
- Added clearClasses method to classStore for resetting class state, improving state management.
- Implemented clear functionality in userStore, projectStore, and workspaceStore to ensure data integrity upon user logout.

simfrisk committed:
Implement workspace user retrieval in UserPage component

- Updated UserPage to fetch users based on the current workspace ID, enhancing user experience by displaying relevant users.
- Added getWorkspaceUsers method in userStore to handle fetching users for a specific workspace, improving data management.
- Refactored useEffect in UserPage to trigger user retrieval when the current workspace ID is available, ensuring accurate user context.

simfrisk committed:
Enhance VideoLibraryPage and related components for improved class creation flow

- Updated VideoLibraryPage to conditionally display the class creation modal based on the current route, enhancing user experience.
- Modified ChooseClassMessage to pass an onClose callback to CreateClass, allowing for better control over modal visibility.
- Refactored CreateClass to handle closing via a callback or state management, improving flexibility in class creation handling.
- Updated SideMenu to only show the "Add new class" button when classes are available, streamlining the UI for teachers.

simfrisk committed:
Refactor modal handling in VideoLibraryPage and CreateClass components

- Updated VideoLibraryPage to include a new ModalContent wrapper, improving event handling by preventing clicks from closing the modal unintentionally.
- Enhanced CreateClass component to stop propagation of click events on the form and button, ensuring better user interaction and modal behavior.

simfrisk committed:
Enhance team management and invitation functionality

- Added a new route for fetching classes accessible by team members, improving team-specific class visibility.
- Introduced a TeamDetailPage for displaying detailed information about teams, enhancing user experience for teachers and students.
- Updated TeamsSection component to navigate to the TeamDetailPage, allowing users to view team-specific details and statistics.
- Implemented fetchInvitationHistory method in workspaceStore to retrieve invitation history, improving invitation management capabilities.

simfrisk committed:
Refactor UserPage and CreateTeam components for improved functionality

- Moved TeamsSection rendering in UserPage to ensure it displays correctly for teachers.
- Updated TeamsSection to use auto-fill for grid layout, enhancing responsiveness.
- Refactored CreateTeam component to improve form handling and user experience, including better state management and UI enhancements for team creation.

simfrisk committed:
Implement invitation management features for teachers

- Added routes for deleting and deactivating invitations, enhancing control over invitation links.
- Updated workspace invitation controller to handle deletion and deactivation logic, ensuring only authorized users can perform these actions.
- Enhanced API documentation for new invitation management endpoints, clarifying usage and permissions.
- Refactored front-end components to integrate new invitation management functionalities, improving user experience for teachers managing invitations.

simfrisk committed:
Enhance team management by adding assignedStudents functionality

- Updated Team model to include assignedStudents field, allowing for better tracking of team members.
- Modified postTeamMember and deleteTeamMember controllers to manage assignedStudents when adding or removing users from teams.
- Enhanced getTeams controller to populate assignedStudents, improving data retrieval for team details.
- Refactored front-end components to display assignedStudents in TeamsSection and TeamDetailPage, ensuring comprehensive visibility of team members.

simfrisk committed:
Enhance team and user management by adding role and profileImage fields

- Updated Team model to include role and profileImage properties for assignedStudents, improving data structure for team members.
- Modified getTeamById and getTeams controllers to populate role and profileImage for assignedTeachers and assignedStudents, enhancing data retrieval.
- Adjusted postUser logic to dynamically assign users to teams based on their role, streamlining team member management during sign-up.

simfrisk committed:
Refactor TeamDetailPage for improved invitation display and user interaction

- Introduced a new state for managing copied invitation IDs, enhancing user feedback when copying links.
- Refactored invitation display components to improve layout and responsiveness, utilizing flexbox for better alignment.
- Updated styles for invitation cards and buttons to enhance visual clarity and user experience.
- Streamlined the invitation actions section, consolidating functionality and improving accessibility on smaller screens.

simfrisk committed:
Update getWorkspaceUsers controller to include profileImage in user retrieval

- Modified the query in getWorkspaceUsers to select the profileImage field along with name, email, and role, enhancing the user data returned for workspaces.

simfrisk committed:
Enhance workspace invitation functionality by adding role validation and UI updates

- Updated createInvitationLink controller to accept an allowedRole parameter, validating it against predefined roles (student, teacher).
- Modified SignUpForm to display the invitation role dynamically based on the invitation token validation response.
- Enhanced handleSignUpSubmit to set the user role based on the invitation role, defaulting to teacher if not specified.
- Introduced a modal in TeachersSection for generating and copying teacher invitation links, improving user experience for managing invitations.

simfrisk committed:
Enhance backend functionality with migration script and documentation update

- Added a new script for database migration in package.json to streamline workspace management.
- Updated API documentation for the validateInvitationToken function to correct formatting and improve clarity on role mismatch responses.

simfrisk committed:
Fix placeholder user creation in migration script

- Updated the migrateToWorkspaces function to correctly assign the created placeholder user from the UserModel.create call, ensuring the placeholder user is properly initialized during migration.

simfrisk committed:
Refactor migration script to ensure session management consistency

- Moved session.endSession() to a finally block in the migrateToWorkspaces function to guarantee that the session is properly closed after both successful and failed migrations, enhancing reliability of the migration process.

simfrisk committed:
Refactor UserPage and TeamDetailPage components to streamline functionality

- Removed unused CreateUserSection import from UserPage, simplifying the component structure.
- Eliminated the handleDeactivateInvitation function from TeamDetailPage, reducing complexity and improving maintainability.
- Updated calendar picker styles in TeamDetailPage to utilize theme-based filtering for better visual consistency.

simfrisk committed:
Enhance Navigation and UserPage components with loading states and workspace management

- Added useEffect in Navigation to fetch user workspaces on mount for logged-in users, improving workspace initialization.
- Introduced loading state in UserPage to display a loading spinner while fetching workspace users, enhancing user experience during data retrieval.
- Refactored UserPage to conditionally render user sections based on loading state, improving responsiveness and clarity.

JennieDalgren commented on frontend/.env.production:
```diff
@@ -0,0 +1 @@
+VITE_API_BASE_URL=https://project-final-simon.onrender.com
```
add the env files to git ignore?

JennieDalgren commented on frontend/public/ bg10.png:
You have quite a lot of images and videos, maybe that's okay for now but be aware that the performance can suffer from this.

JennieDalgren commented on backend/dist/controllers/OLD-postProject.js:
```diff
@@ -0,0 +1,31 @@
+"use strict";
```
is this a file that's not in use?
If so, remove it?

JennieDalgren reviewed (approved):
Great job with this project!
You've done a lot of things and it is approved with VG

Some things you nailed:
• Your backend code is structured well with clear separation of concerns through different controllers.
• The use of async/await along with proper error handling is great.
• Your use of Swagger for API documentation is spot-on. it’s easy to follow and understand.
• Nice looking UI and polished product!
• Great documentation of your work, how to use the project, demo users etc.

Remember to clean up unused code and files.
Also, why not add a favicon to the site! :)

simfrisk committed:
Update README.md

simfrisk committed:
Update README to clarify backend section heading

simfrisk committed:
Test GPG signature

simfrisk committed:
test

simfrisk committed:
test

simfrisk committed:
message

simfrisk committed:
test

simfrisk committed:
Update routes, workspace controllers, and review page components

simfrisk committed:
Add timeline zoom functionality to VideoSection component

simfrisk committed:
refactor MarkerMessage component in VideoSection to dynamically adjust max-width based on position, preventing text clipping near edges

simfrisk committed:
Add subtitle functionality to VideoSection component, allowing users to toggle subtitles and display relevant comments based on video playback time.

simfrisk committed:
Update invitation messages in SignUpForm component to include localized text for non-teacher roles

simfrisk committed:
invitation message in SignUpForm component for student role

simfrisk committed:
Add health check endpoint and backend status indicator components

- Introduced a new health check endpoint in the backend to monitor API status.
- Added BackendStatusIndicator component to display backend connection status in the frontend.
- Integrated health check functionality into user login and sign-up processes to handle cold start scenarios.
- Updated existing components to include the BackendStatusIndicator for improved user experience during backend initialization.

simfrisk committed:
Remove unused index on token field in WorkspaceInvitation schema for improved performance.
